WNPRC_Purchasing/src/client/components/LineItemsPanel.tsx

```diff
@@ -50,67 +53,94 @@ export const LineItemsPanel: FC<Props> = memo(props => {
+                                <Col xs={4}>Part no./Item description *</Col>
+                                <Col xs={1}>Unit *</Col>
+                                { (isAdmin || !canUpdate) &&
+                                    <Col xs={1}>Unit Cost ($) *</Col>
+                                }
+                                <Col xs={1}>Quantity *</Col>
+                                { (isAdmin || !canUpdate) &&
+                                    <Col xs={1}>Subtotal ($)</Col>
+                                }
+                                <Col xs={!hasRequestId ? 2 : 1}>Controlled substance</Col>
+                                { hasRequestId && (isAdmin || canUpdate) &&
+                                    <Col xs={1}>Quantity received</Col>
+                                }
+                            </Row>
+                        }
+                    </div>
+                    <div>
+                        {lineItems.map((lineItem, idx) => {
+                            return (
+                                <LineItemRow
+                                    rowIndex={idx}
+                                    key={'line-item-' + idx}
+                                    model={lineItem}
+                                    onInputChange={lineItemRowChange}
+                                    onDelete={onDeleteRow}
+                                    isAdmin={isAdmin}
+                                    canUpdate={canUpdate}
+                                    canInsert={canInsert}
+                                    hasRequestId={hasRequestId}
+                                />
+                            );
+                        })}
+                    </div>
+                    <div>
+                        {
+                            (isAdmin || !canUpdate) && (
+                                <>
+                                    <div>
+                                        <Row>
+                                            <Col xs={4}></Col>
+                                            <Col xs={1}></Col>
+                                            <Col xs={1}></Col>
+                                            <Col className="calc-total" xs={1}>
+                                                Total ($):
+                                            </Col>
+                                            <Col className="calc-total-val" xs={1}>
+                                                {formatCurrency(
+                                                    lineItems.reduce(function (accumulatedVal, lineItem) {
+                                                        return lineItem.unitCost && lineItem.quantity
+                                                            ? accumulatedVal + lineItem.unitCost * lineItem.quantity
+                                                            : accumulatedVal;
+                                                    }, 0)
```

Comment:
I would pull this out of the JSX and make it a useMemo.  It's a lot of JS in JSX (which is a messy language to begin with), also it's getting recalculated each render.

Reply:
done